File: chatdownloader/src/_types/twitchtypes.rs
/*
Contains all the Twitch types parsable from the chat log
*/

use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug)]
pub struct ChatMessageFragmentEmoticon {
    /* 
    Represents an emoticon in a chat message fragment.
    */
    pub emoticon_id: String,
}

#[derive(Deserialize, Serialize, Debug)]
pub struct ChatMessageFragment {
    /*
    Represents a fragment of a chat message.
    */
    pub text: String,
    pub emoticon: Option<ChatMessageFragmentEmoticon>,
}

#[derive(Deserialize, Serialize, Debug)]
pub struct Badge {
    /*
    Represents a badge.
    */
    pub _id: String,
    pub version: String,
}

#[derive(Deserialize, Serialize, Debug)]
pub struct ChatMessage {
    /*
    Represents a chat message.
    */
    pub body: String,
    pub bits_spent: Option<u32>,
    pub fragments: Vec<ChatMessageFragment>,
    pub user_badges: Option<Vec<Badge>>,
}

#[derive(Deserialize, Serialize, Debug)]
pub struct ChatUserInfo {
    /*
    Represents a user in a chat.
    */
    pub display_name: String,
    pub _id: String,
    pub logo: String,
}

#[derive(Deserialize, Serialize, Debug)]
pub struct Comment {
    /*
    Represents a comment in a chat.
    */
    pub _id: String,
    pub message: ChatMessage,
    pub commenter: ChatUserInfo,
}

#[derive(Deserialize, Serialize, Debug)]
pub struct ChatLog {
    /*
    Represents a chat log.
    */
    pub comments: Vec<Comment>,
}

File: python-chatdownloader/src/_types/twitchtypes.py
"""
Contains all the Twitch types parsable from the chat log
"""

from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from dataclasses_json import DataClassJsonMixin


@dataclass
class ChatMessageFragmentEmoticon(DataClassJsonMixin):
    """
    Represents an emoticon in a chat message fragment.
    """
    emoticon_id: str


@dataclass
class ChatMessageFragment(DataClassJsonMixin):
    """
    Represents a fragment of a chat message.
    """
    text: str
    emoticon: Optional[ChatMessageFragmentEmoticon]


@dataclass
class Badge(DataClassJsonMixin):
    """
    Represents a badge.
    """
    _id: str
    version: str


@dataclass
class ChatMessage(DataClassJsonMixin):
    """
    Represents a chat message.
    """
    body: str
    bits_spent: int
    fragments: List[ChatMessageFragment]
    user_badges: List[Badge]


@dataclass
class ChatUserInfo(DataClassJsonMixin):
    """
    Represents a user in a chat.
    """
    display_name: str
    _id: str
    logo: str


@dataclass
class Comment(DataClassJsonMixin):
    """
    Represents a comment in a chat.
    """
    _id: str
    message: ChatMessage
    commenter: ChatUserInfo


@dataclass
class ChatLog(DataClassJsonMixin):
    """
    Represents a chat log.
    """
    comments: List[Comment]
